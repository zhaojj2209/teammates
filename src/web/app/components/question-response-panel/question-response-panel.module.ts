import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { InViewportModule } from 'ng-in-viewport';
import { LoadingRetryModule } from '../loading-retry/loading-retry.module';
import { LoadingSpinnerModule } from '../loading-spinner/loading-spinner.module';
import { SingleStatisticsModule } from '../question-responses/single-statistics/single-statistics.module';
import {
  StudentViewResponsesModule,
} from '../question-responses/student-view-responses/student-view-responses.module';
import { QuestionTextWithInfoModule } from '../question-text-with-info/question-text-with-info.module';

import { QuestionResponsePanelComponent } from './question-response-panel.component';

/**
 * Question Response Panel module.
 */
@NgModule({
  imports: [
    CommonModule,
    QuestionTextWithInfoModule,
    SingleStatisticsModule,
    StudentViewResponsesModule,
    InViewportModule,
    LoadingSpinnerModule,
    LoadingRetryModule,
  ],
  declarations: [
    QuestionResponsePanelComponent,
  ],
  exports: [
    QuestionResponsePanelComponent,
  ],
})
export class QuestionResponsePanelModule { }
