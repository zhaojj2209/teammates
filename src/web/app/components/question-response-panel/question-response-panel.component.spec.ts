import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { of } from 'rxjs';
import SpyInstance = jest.SpyInstance;
import { FeedbackSessionsService } from '../../../services/feedback-sessions.service';
import {
  FeedbackContributionQuestionDetails,
  FeedbackContributionResponseDetails, FeedbackMcqQuestionDetails,
  FeedbackMcqResponseDetails, FeedbackParticipantType,
  FeedbackQuestion, FeedbackQuestionType,
  FeedbackRubricQuestionDetails, FeedbackRubricResponseDetails,
  FeedbackSession, FeedbackSessionPublishStatus, FeedbackSessionSubmissionStatus,
  FeedbackTextQuestionDetails, FeedbackTextResponseDetails, FeedbackVisibilityType,
  NumberOfEntitiesToGiveFeedbackToSetting, ResponseVisibleSetting,
  SessionResults, SessionVisibleSetting,
} from '../../../types/api-output';
import { FeedbackQuestionModel } from '../../pages-session/session-result-page/session-result-page.component';
import { LoadingRetryModule } from '../loading-retry/loading-retry.module';
import { LoadingSpinnerModule } from '../loading-spinner/loading-spinner.module';
import { SingleStatisticsModule } from '../question-responses/single-statistics/single-statistics.module';
import {
  StudentViewResponsesModule,
} from '../question-responses/student-view-responses/student-view-responses.module';
import { QuestionTextWithInfoModule } from '../question-text-with-info/question-text-with-info.module';
import { QuestionResponsePanelComponent } from './question-response-panel.component';

describe('QuestionResponsePanelComponent', () => {

  const testFeedbackSession: FeedbackSession = {
    feedbackSessionName: 'First Session',
    courseId: 'CS1231',
    timeZone: 'Asia/Singapore',
    instructions: '',
    submissionStartTimestamp: 0,
    submissionEndTimestamp: 1549095330000,
    gracePeriod: 0,
    sessionVisibleSetting: SessionVisibleSetting.AT_OPEN,
    responseVisibleSetting: ResponseVisibleSetting.AT_VISIBLE,
    submissionStatus: FeedbackSessionSubmissionStatus.OPEN,
    publishStatus: FeedbackSessionPublishStatus.PUBLISHED,
    isClosingEmailEnabled: true,
    isPublishedEmailEnabled: true,
    createdAtTimestamp: 0,
    studentDeadlines: {},
    instructorDeadlines: {},
  };

  const testQuestion1: FeedbackQuestion = {
    feedbackQuestionId: 'feedbackQuestion1',
    questionNumber: 1,
    questionBrief: 'How well did team member perform?',
    questionDescription: '',
    questionDetails: {
      hasAssignedWeights: false,
      mcqWeights: [],
      mcqOtherWeight: 0,
      mcqChoices: [
        '<p>Good</p>',
        '<p>Normal</p>',
        '<p>Bad</p>',
      ],
      otherEnabled: false,
      questionDropdownEnabled: false,
      generateOptionsFor: 'NONE',
      questionType: FeedbackQuestionType.MCQ,
      questionText: 'How well did team member perform?',
    } as FeedbackMcqQuestionDetails,
    questionType: FeedbackQuestionType.MCQ,
    giverType: FeedbackParticipantType.STUDENTS,
    recipientType: FeedbackParticipantType.OWN_TEAM_MEMBERS_INCLUDING_SELF,
    numberOfEntitiesToGiveFeedbackToSetting: NumberOfEntitiesToGiveFeedbackToSetting.UNLIMITED,
    showResponsesTo: [],
    showGiverNameTo: [],
    showRecipientNameTo: [],
    customNumberOfEntitiesToGiveFeedbackTo: 0,
  };

  const testQuestion2: FeedbackQuestion = {
    feedbackQuestionId: 'feedbackQuestion2',
    questionNumber: 2,
    questionBrief: 'Rate your teammates in contribution',
    questionDescription: '',
    questionDetails: {
      questionType: FeedbackQuestionType.CONTRIB,
      questionText: 'Rate your teammates in contribution',
      isNotSureAllowed: false,
    } as FeedbackContributionQuestionDetails,
    questionType: FeedbackQuestionType.CONTRIB,
    giverType: FeedbackParticipantType.STUDENTS,
    recipientType: FeedbackParticipantType.OWN_TEAM_MEMBERS,
    numberOfEntitiesToGiveFeedbackToSetting: NumberOfEntitiesToGiveFeedbackToSetting.UNLIMITED,
    showResponsesTo: [],
    showGiverNameTo: [],
    showRecipientNameTo: [],
    customNumberOfEntitiesToGiveFeedbackTo: 0,
  };

  const testQuestion3: FeedbackQuestion = {
    feedbackQuestionId: 'feedbackQuestion3',
    questionNumber: 3,
    questionBrief: 'Rate your teammates proficiency',
    questionDescription: '',
    questionDetails: {
      questionType: FeedbackQuestionType.RUBRIC,
      questionText: 'Rate your teammates proficiency',
      hasAssignedWeights: false,
      rubricWeightsForEachCell: [[]],
      rubricChoices: ['Poor', 'Average', 'Good'],
      rubricSubQuestions: [],
      rubricDescriptions: [[]],
    } as FeedbackRubricQuestionDetails,
    questionType: FeedbackQuestionType.RUBRIC,
    giverType: FeedbackParticipantType.STUDENTS,
    recipientType: FeedbackParticipantType.OWN_TEAM_MEMBERS,
    numberOfEntitiesToGiveFeedbackToSetting: NumberOfEntitiesToGiveFeedbackToSetting.UNLIMITED,
    showResponsesTo: [],
    showGiverNameTo: [],
    showRecipientNameTo: [],
    customNumberOfEntitiesToGiveFeedbackTo: 0,
  };

  const testQuestionAnonymousResponse1: FeedbackQuestion = {
    feedbackQuestionId: 'feedbackQuestionAnonymousResponse1',
    questionNumber: 1,
    questionBrief: 'What comments do you have regarding each of your team members? '
        + '(response is confidential and will only be shown to the instructor).',
    questionDescription: '',
    questionDetails: {
      shouldAllowRichText: true,
      questionType: FeedbackQuestionType.TEXT,
      questionText: 'What comments do you have regarding each of your team members? '
          + '(response is confidential and will only be shown to the instructor).',
    } as FeedbackTextQuestionDetails,
    questionType: FeedbackQuestionType.TEXT,
    giverType: FeedbackParticipantType.STUDENTS,
    recipientType: FeedbackParticipantType.OWN_TEAM_MEMBERS,
    numberOfEntitiesToGiveFeedbackToSetting: NumberOfEntitiesToGiveFeedbackToSetting.UNLIMITED,
    showResponsesTo: [FeedbackVisibilityType.INSTRUCTORS],
    showGiverNameTo: [FeedbackVisibilityType.INSTRUCTORS],
    showRecipientNameTo: [FeedbackVisibilityType.INSTRUCTORS],
    customNumberOfEntitiesToGiveFeedbackTo: 0,
  };

  const testQuestionAnonymousResponse2: FeedbackQuestion = {
    feedbackQuestionId: 'feedbackQuestionAnonymousResponse2',
    questionNumber: 2,
    questionBrief: 'How are the team dynamics thus far? '
        + '(response is confidential and will only be shown to the instructor).',
    questionDescription: '',
    questionDetails: {
      shouldAllowRichText: true,
      questionType: FeedbackQuestionType.TEXT,
      questionText: 'How are the team dynamics thus far? '
          + '(response is confidential and will only be shown to the instructor).',
    } as FeedbackTextQuestionDetails,
    questionType: FeedbackQuestionType.TEXT,
    giverType: FeedbackParticipantType.STUDENTS,
    recipientType: FeedbackParticipantType.OWN_TEAM,
    numberOfEntitiesToGiveFeedbackToSetting: NumberOfEntitiesToGiveFeedbackToSetting.UNLIMITED,
    showResponsesTo: [FeedbackVisibilityType.INSTRUCTORS],
    showGiverNameTo: [FeedbackVisibilityType.INSTRUCTORS],
    showRecipientNameTo: [FeedbackVisibilityType.INSTRUCTORS],
    customNumberOfEntitiesToGiveFeedbackTo: 0,
  };

  const testFeedbackQuestionModel: FeedbackQuestionModel = {
    feedbackQuestion: testQuestion1,
    questionStatistics: '',
    allResponses: [],
    responsesToSelf: [],
    responsesFromSelf: [],
    otherResponses: [],
    isLoading: false,
    isLoaded: false,
    hasResponse: true,
  };

  let component: QuestionResponsePanelComponent;
  let fixture: ComponentFixture<QuestionResponsePanelComponent>;
  let feedbackSessionsService: FeedbackSessionsService;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [
        QuestionResponsePanelComponent,
      ],
      imports: [
        HttpClientTestingModule,
        RouterTestingModule,
        SingleStatisticsModule,
        StudentViewResponsesModule,
        QuestionTextWithInfoModule,
        LoadingSpinnerModule,
        LoadingRetryModule,
      ],
      providers: [FeedbackSessionsService],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(QuestionResponsePanelComponent);
    feedbackSessionsService = TestBed.inject(FeedbackSessionsService);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with feedback session with questions', () => {
    component.session = testFeedbackSession;
    component.questions = [
      {
        feedbackQuestion: testQuestion1,
        questionStatistics: '',
        allResponses: [],
        responsesToSelf: [],
        responsesFromSelf: [
          {
            isMissingResponse: false,
            responseId: 'resp-id-1',
            giver: 'giver1',
            giverTeam: 'team1',
            giverSection: 'section1',
            recipient: 'recipient1',
            recipientTeam: 'team1',
            recipientSection: 'section1',
            responseDetails: {
              answer: 'Good',
              isOther: false,
              otherFieldContent: '',
            } as FeedbackMcqResponseDetails,
            instructorComments: [],
          },
        ],
        otherResponses: [[]],
        isLoading: false,
        isLoaded: true,
        hasResponse: true,
      },
      {
        feedbackQuestion: testQuestion2,
        questionStatistics: '',
        allResponses: [],
        responsesToSelf: [
          {
            isMissingResponse: false,
            responseId: 'resp-id-2',
            giver: 'giver1',
            giverTeam: 'team1',
            giverSection: 'section1',
            recipient: 'giver1',
            recipientTeam: 'team1',
            recipientSection: 'section1',
            responseDetails: {
              answer: 120,
            } as FeedbackContributionResponseDetails,
            instructorComments: [],
          },
        ],
        responsesFromSelf: [
          {
            isMissingResponse: false,
            responseId: 'resp-id-3',
            giver: 'giver1',
            giverTeam: 'team1',
            giverSection: 'section1',
            recipient: 'recipient2',
            recipientTeam: 'team2',
            recipientSection: 'section2',
            responseDetails: {
              answer: 110,
            } as FeedbackContributionResponseDetails,
            instructorComments: [],
          },
          {
            isMissingResponse: false,
            responseId: 'resp-id-4',
            giver: 'giver1',
            giverTeam: 'team1',
            giverSection: 'section1',
            recipient: 'recipient3',
            recipientTeam: 'team2',
            recipientSection: 'section2',
            responseDetails: {
              answer: 100,
            } as FeedbackContributionResponseDetails,
            instructorComments: [],
          },
        ],
        otherResponses: [[]],
        isLoading: false,
        isLoaded: true,
        hasResponse: true,
      },
      {
        feedbackQuestion: testQuestion3,
        questionStatistics: '',
        allResponses: [],
        responsesToSelf: [],
        responsesFromSelf: [
          {
            isMissingResponse: false,
            responseId: 'resp-id-5',
            giver: 'giver1',
            giverTeam: 'team1',
            giverSection: 'section1',
            recipient: 'recipient3',
            recipientTeam: 'team2',
            recipientSection: 'section2',
            responseDetails: {
              answer: [1],
            } as FeedbackRubricResponseDetails,
            instructorComments: [
              {
                commentGiver: 'comment-giver-1',
                lastEditorEmail: 'comment@egeg.com',
                feedbackResponseCommentId: 1,
                commentText: 'this is a text',
                createdAt: 1402775804,
                lastEditedAt: 1402775804,
                isVisibilityFollowingFeedbackQuestion: true,
                showGiverNameTo: [],
                showCommentTo: [],
              },
            ],
          },
        ],
        otherResponses: [[]],
        isLoading: false,
        isLoaded: true,
        hasResponse: true,
      },
    ];

    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with feedback session with questions of anonymous responses', () => {
    component.session = testFeedbackSession;
    component.questions = [
      {
        feedbackQuestion: testQuestionAnonymousResponse1,
        questionStatistics: '',
        allResponses: [
          {
            isMissingResponse: false,
            responseId: 'resp-id-1',
            giver: 'giver1',
            giverTeam: 'team1',
            giverSection: 'section1',
            recipient: 'recipient1',
            recipientTeam: 'team1',
            recipientSection: 'section1',
            responseDetails: {
              answer: '<p>asdf</p>',
            } as FeedbackTextResponseDetails,
            instructorComments: [],
          },
          {
            isMissingResponse: false,
            responseId: 'resp-id-2',
            giver: 'giver1',
            giverTeam: 'team1',
            giverSection: 'section1',
            recipient: 'recipient2',
            recipientTeam: 'team1',
            recipientSection: 'section1',
            responseDetails: {
              answer: '<p>asdf</p>',
            } as FeedbackTextResponseDetails,
            instructorComments: [],
          },
          {
            isMissingResponse: false,
            responseId: 'resp-id-3',
            giver: 'giver1',
            giverTeam: 'team1',
            giverSection: 'section1',
            recipient: 'recipient3',
            recipientTeam: 'team1',
            recipientSection: 'section1',
            responseDetails: {
              answer: '<p>asdf</p>',
            } as FeedbackTextResponseDetails,
            instructorComments: [],
          },
        ],
        responsesToSelf: [],
        responsesFromSelf: [
          {
            isMissingResponse: false,
            responseId: 'resp-id-1',
            giver: 'giver1',
            giverTeam: 'team1',
            giverSection: 'section1',
            recipient: 'recipient1',
            recipientTeam: 'team1',
            recipientSection: 'section1',
            responseDetails: {
              answer: '<p>asdf</p>',
            } as FeedbackTextResponseDetails,
            instructorComments: [],
          },
          {
            isMissingResponse: false,
            responseId: 'resp-id-2',
            giver: 'giver1',
            giverTeam: 'team1',
            giverSection: 'section1',
            recipient: 'recipient2',
            recipientTeam: 'team1',
            recipientSection: 'section1',
            responseDetails: {
              answer: '<p>asdf</p>',
            } as FeedbackTextResponseDetails,
            instructorComments: [],
          },
          {
            isMissingResponse: false,
            responseId: 'resp-id-3',
            giver: 'giver1',
            giverTeam: 'team1',
            giverSection: 'section1',
            recipient: 'recipient3',
            recipientTeam: 'team1',
            recipientSection: 'section1',
            responseDetails: {
              answer: '<p>asdf</p>',
            } as FeedbackTextResponseDetails,
            instructorComments: [],
          },
        ],
        otherResponses: [],
        isLoading: false,
        isLoaded: true,
        hasResponse: true,
      },
      {
        feedbackQuestion: testQuestionAnonymousResponse2,
        questionStatistics: '',
        allResponses: [
          {
            isMissingResponse: false,
            responseId: 'resp-id-4',
            giver: 'giver1',
            giverTeam: 'team1',
            giverSection: 'section1',
            recipient: 'team1',
            recipientTeam: 'team1',
            recipientSection: 'section1',
            responseDetails: {
              answer: '<p>asdf</p>',
            } as FeedbackTextResponseDetails,
            instructorComments: [],
          },
        ],
        responsesToSelf: [],
        responsesFromSelf: [
          {
            isMissingResponse: false,
            responseId: 'resp-id-4',
            giver: 'giver1',
            giverTeam: 'team1',
            giverSection: 'section1',
            recipient: 'team1',
            recipientTeam: 'team1',
            recipientSection: 'section1',
            responseDetails: {
              answer: '<p>asdf</p>',
            } as FeedbackTextResponseDetails,
            instructorComments: [],
          },
        ],
        otherResponses: [],
        isLoading: false,
        isLoaded: true,
        hasResponse: true,
      },
    ];

    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should load the recipients and responses of a question if not yet loaded', () => {
    component.session = testFeedbackSession;
    component.questions = [testFeedbackQuestionModel];
    const testFeedbackSessionResult: SessionResults = {
      questions: [
        {
          feedbackQuestion: testQuestion1,
          questionStatistics: '',
          allResponses: [],
          responsesToSelf: [],
          responsesFromSelf: [],
          otherResponses: [],
        },
      ],
    };

    const fsSpy: SpyInstance = jest.spyOn(feedbackSessionsService, 'getFeedbackSessionResults')
        .mockReturnValue(of(testFeedbackSessionResult));
    component.loadQuestion({ visible: true }, testFeedbackQuestionModel);

    expect(fsSpy).toBeCalledTimes(1);
    expect(fsSpy).toHaveBeenLastCalledWith({
      intent: 'STUDENT_RESULT',
      courseId: 'CS1231',
      feedbackSessionName: 'First Session',
      questionId: testQuestion1.feedbackQuestionId,
      key: '',
    });
    expect(testFeedbackQuestionModel.isLoading).toBe(false);
    expect(testFeedbackQuestionModel.isLoaded).toBe(true);
    expect(testFeedbackQuestionModel.hasResponse).toBe(true);
  });

  it('should not load the recipients and responses of a question if already loaded', () => {
    const fsSpy: SpyInstance = jest.spyOn(feedbackSessionsService, 'getFeedbackSessionResults');

    testFeedbackQuestionModel.isLoaded = true;
    component.loadQuestion({ visible: true }, testFeedbackQuestionModel);

    testFeedbackQuestionModel.isLoading = true;
    component.loadQuestion({ visible: true }, testFeedbackQuestionModel);

    expect(fsSpy).not.toHaveBeenCalled();
  });
});
