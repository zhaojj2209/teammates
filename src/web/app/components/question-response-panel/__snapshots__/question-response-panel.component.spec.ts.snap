// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionResponsePanelComponent should snap with feedback session with questions 1`] = `
<tm-question-response-panel
  RESPONSE_HIDDEN_QUESTIONS={[Function Array]}
  feedbackSessionsService={[Function FeedbackSessionsService]}
  intent={[Function String]}
  questions={[Function Array]}
  regKey=""
  session={[Function Object]}
  statusMessageService={[Function StatusMessageService]}
>
  <div
    inviewport=""
  >
    <div />
    <tm-loading-retry>
      <div
        class="card bg-light mt-4"
        id="question-1-responses"
      >
        <div
          class="card-body"
        >
          <tm-question-text-with-info
            class="question-text"
          >
            <span
              class="d-inline-flex row align-items-center no-gutters"
            >
              <span
                class="col-auto"
              >
                <b>
                  Question 1: 
                </b>
              </span>
              <span
                class="col-auto"
                id="question-text"
              >
                 How well did team member perform?
              </span>
              <a
                class="info-font-size ml-1"
                href="/"
                id="additional-info-link"
                queryparamshandling="merge"
              >
                 [more] 
              </a>
            </span>
          </tm-question-text-with-info>
          <div
            class="other-response"
          >
            <strong>
              Other responses (to you): 
            </strong>
            Responses are not visible to you. 
          </div>
          <div
            class="mt-4"
            id="given-responses"
          >
            <strong>
              Your own responses (to others):
            </strong>
            <div>
              <tm-student-view-responses>
                <div
                  class="card border-info mt-4"
                >
                  <div
                    class="card-header bg-info text-white"
                  >
                    <span
                      id="response-recipient"
                    >
                      <b>
                        To:
                      </b>
                       recipient1
                    </span>
                  </div>
                  <div
                    class="card-body"
                    style="padding: 0;"
                  >
                    <table
                      class="table"
                    >
                      <tbody>
                        <tr>
                          <td
                            id="response-giver"
                          >
                            <strong>
                              From:
                            </strong>
                             giver1 
                          </td>
                        </tr>
                        <tr>
                          <td>
                            <tm-single-response>
                            </tm-single-response>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </tm-student-view-responses>
            </div>
          </div>
        </div>
      </div>
    </tm-loading-retry>
  </div><div
    inviewport=""
  >
    <div />
    <tm-loading-retry>
      <div
        class="card bg-light mt-4"
        id="question-2-responses"
      >
        <div
          class="card-body"
        >
          <tm-question-text-with-info
            class="question-text"
          >
            <span
              class="d-inline-flex row align-items-center no-gutters"
            >
              <span
                class="col-auto"
              >
                <b>
                  Question 2: 
                </b>
              </span>
              <span
                class="col-auto"
                id="question-text"
              >
                 Rate your teammates in contribution
              </span>
              <a
                class="info-font-size ml-1"
                href="/"
                id="additional-info-link"
                queryparamshandling="merge"
              >
                 [more] 
              </a>
            </span>
          </tm-question-text-with-info>
          <div
            class="other-response"
          >
            <strong>
              Other responses (to you): 
            </strong>
            Responses are not visible to you. 
          </div>
          <div
            class="mt-4"
            id="given-responses"
          >
            <strong>
              Your own responses (to others):
            </strong>
            <div>
              <tm-student-view-responses>
                <div
                  class="card border-info mt-4"
                >
                  <div
                    class="card-header bg-info text-white"
                  >
                    <span
                      id="response-recipient"
                    >
                      <b>
                        To:
                      </b>
                       recipient2
                    </span>
                  </div>
                  <div
                    class="card-body"
                    style="padding: 0;"
                  >
                    <table
                      class="table"
                    >
                      <tbody>
                        <tr>
                          <td
                            id="response-giver"
                          >
                            <strong>
                              From:
                            </strong>
                             giver1 
                          </td>
                        </tr>
                        <tr>
                          <td>
                            <tm-single-response>
                            </tm-single-response>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </tm-student-view-responses>
            </div>
            <div>
              <tm-student-view-responses>
                <div
                  class="card border-info mt-4"
                >
                  <div
                    class="card-header bg-info text-white"
                  >
                    <span
                      id="response-recipient"
                    >
                      <b>
                        To:
                      </b>
                       recipient3
                    </span>
                  </div>
                  <div
                    class="card-body"
                    style="padding: 0;"
                  >
                    <table
                      class="table"
                    >
                      <tbody>
                        <tr>
                          <td
                            id="response-giver"
                          >
                            <strong>
                              From:
                            </strong>
                             giver1 
                          </td>
                        </tr>
                        <tr>
                          <td>
                            <tm-single-response>
                            </tm-single-response>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </tm-student-view-responses>
            </div>
          </div>
        </div>
      </div>
    </tm-loading-retry>
  </div><div
    inviewport=""
  >
    <div />
    <tm-loading-retry>
      <div
        class="card bg-light mt-4"
        id="question-3-responses"
      >
        <div
          class="card-body"
        >
          <tm-question-text-with-info
            class="question-text"
          >
            <span
              class="d-inline-flex row align-items-center no-gutters"
            >
              <span
                class="col-auto"
              >
                <b>
                  Question 3: 
                </b>
              </span>
              <span
                class="col-auto"
                id="question-text"
              >
                 Rate your teammates proficiency
              </span>
              <a
                class="info-font-size ml-1"
                href="/"
                id="additional-info-link"
                queryparamshandling="merge"
              >
                 [more] 
              </a>
            </span>
          </tm-question-text-with-info>
          <div
            class="other-response"
          >
            <strong>
              Other responses (to you): 
            </strong>
            Responses are not visible to you. 
          </div>
          <div
            class="mt-4"
            id="given-responses"
          >
            <strong>
              Your own responses (to others):
            </strong>
            <div>
              <tm-student-view-responses>
                <div
                  class="card border-info mt-4"
                >
                  <div
                    class="card-header bg-info text-white"
                  >
                    <span
                      id="response-recipient"
                    >
                      <b>
                        To:
                      </b>
                       recipient3
                    </span>
                  </div>
                  <div
                    class="card-body"
                    style="padding: 0;"
                  >
                    <table
                      class="table"
                    >
                      <tbody>
                        <tr>
                          <td
                            id="response-giver"
                          >
                            <strong>
                              From:
                            </strong>
                             giver1 
                          </td>
                        </tr>
                        <tr>
                          <td>
                            <tm-single-response>
                            </tm-single-response>
                          </td>
                        </tr>
                        <tr>
                          <td>
                            <tm-comment-table>
                              <div
                                class="card"
                              >
                                <ul
                                  class="list-group list-group-flush"
                                >
                                  <li
                                    class="list-group-item"
                                  >
                                    <tm-comment-row>
                                      <div
                                        class="card"
                                      >
                                        <div
                                          class="card-body"
                                        >
                                          <div
                                            class="row comment-row"
                                          >
                                            <div
                                              class="col-12"
                                            >
                                              <span
                                                class="text-secondary"
                                              >
                                                <span
                                                  id="comment-giver-name"
                                                >
                                                  comment-giver-1 commented at 
                                                </span>
                                                <span
                                                  class="ngb-tooltip-class"
                                                  style="margin-right: .25rem;"
                                                >
                                                   17 Jan 1:09 PM
                                                </span>
                                              </span>
                                              <i
                                                aria-hidden="true"
                                                class="fa fa-eye"
                                              />
                                              <div
                                                class="float-right"
                                              >
                                              </div>
                                            </div>
                                            <div
                                              class="col-12"
                                              id="comment-text"
                                            >
                                              this is a text
                                            </div>
                                          </div>
                                        </div>
                                      </div>
                                    </tm-comment-row>
                                  </li>
                                </ul>
                              </div>
                            </tm-comment-table>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </tm-student-view-responses>
            </div>
          </div>
        </div>
      </div>
    </tm-loading-retry>
  </div>
</tm-question-response-panel>
`;

exports[`QuestionResponsePanelComponent should snap with feedback session with questions of anonymous responses 1`] = `
<tm-question-response-panel
  RESPONSE_HIDDEN_QUESTIONS={[Function Array]}
  feedbackSessionsService={[Function FeedbackSessionsService]}
  intent={[Function String]}
  questions={[Function Array]}
  regKey=""
  session={[Function Object]}
  statusMessageService={[Function StatusMessageService]}
>
  <div
    inviewport=""
  >
    <div />
    <tm-loading-retry>
      <div
        class="card bg-light mt-4"
        id="question-1-responses"
      >
        <div
          class="card-body"
        >
          <tm-question-text-with-info
            class="question-text"
          >
            <span
              class="d-inline-flex row align-items-center no-gutters"
            >
              <span
                class="col-auto"
              >
                <b>
                  Question 1: 
                </b>
              </span>
              <span
                class="col-auto"
                id="question-text"
              >
                 What comments do you have regarding each of your team members? (response is confidential and will only be shown to the instructor).
              </span>
            </span>
          </tm-question-text-with-info>
          <div
            class="other-response"
          >
            <strong>
              Other responses (to you): 
            </strong>
            Responses are not visible to you. 
          </div>
          <div
            class="mt-4"
            id="given-responses"
          >
            <strong>
              Your own responses (to others):
            </strong>
            <div>
              <tm-student-view-responses>
                <div
                  class="card border-info mt-4"
                >
                  <div
                    class="card-header bg-info text-white"
                  >
                    <span
                      id="response-recipient"
                    >
                      <b>
                        To:
                      </b>
                       recipient1
                    </span>
                  </div>
                  <div
                    class="card-body"
                    style="padding: 0;"
                  >
                    <table
                      class="table"
                    >
                      <tbody>
                        <tr>
                          <td
                            id="response-giver"
                          >
                            <strong>
                              From:
                            </strong>
                             giver1 
                          </td>
                        </tr>
                        <tr>
                          <td>
                            <tm-single-response>
                            </tm-single-response>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </tm-student-view-responses>
            </div>
            <div>
              <tm-student-view-responses>
                <div
                  class="card border-info mt-4"
                >
                  <div
                    class="card-header bg-info text-white"
                  >
                    <span
                      id="response-recipient"
                    >
                      <b>
                        To:
                      </b>
                       recipient2
                    </span>
                  </div>
                  <div
                    class="card-body"
                    style="padding: 0;"
                  >
                    <table
                      class="table"
                    >
                      <tbody>
                        <tr>
                          <td
                            id="response-giver"
                          >
                            <strong>
                              From:
                            </strong>
                             giver1 
                          </td>
                        </tr>
                        <tr>
                          <td>
                            <tm-single-response>
                            </tm-single-response>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </tm-student-view-responses>
            </div>
            <div>
              <tm-student-view-responses>
                <div
                  class="card border-info mt-4"
                >
                  <div
                    class="card-header bg-info text-white"
                  >
                    <span
                      id="response-recipient"
                    >
                      <b>
                        To:
                      </b>
                       recipient3
                    </span>
                  </div>
                  <div
                    class="card-body"
                    style="padding: 0;"
                  >
                    <table
                      class="table"
                    >
                      <tbody>
                        <tr>
                          <td
                            id="response-giver"
                          >
                            <strong>
                              From:
                            </strong>
                             giver1 
                          </td>
                        </tr>
                        <tr>
                          <td>
                            <tm-single-response>
                            </tm-single-response>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </tm-student-view-responses>
            </div>
          </div>
        </div>
      </div>
    </tm-loading-retry>
  </div><div
    inviewport=""
  >
    <div />
    <tm-loading-retry>
      <div
        class="card bg-light mt-4"
        id="question-2-responses"
      >
        <div
          class="card-body"
        >
          <tm-question-text-with-info
            class="question-text"
          >
            <span
              class="d-inline-flex row align-items-center no-gutters"
            >
              <span
                class="col-auto"
              >
                <b>
                  Question 2: 
                </b>
              </span>
              <span
                class="col-auto"
                id="question-text"
              >
                 How are the team dynamics thus far? (response is confidential and will only be shown to the instructor).
              </span>
            </span>
          </tm-question-text-with-info>
          <div
            class="other-response"
          >
            <strong>
              Other responses (to you): 
            </strong>
            Responses are not visible to you. 
          </div>
          <div
            class="mt-4"
            id="given-responses"
          >
            <strong>
              Your own responses (to others):
            </strong>
            <div>
              <tm-student-view-responses>
                <div
                  class="card border-info mt-4"
                >
                  <div
                    class="card-header bg-info text-white"
                  >
                    <span
                      id="response-recipient"
                    >
                      <b>
                        To:
                      </b>
                       team1
                    </span>
                  </div>
                  <div
                    class="card-body"
                    style="padding: 0;"
                  >
                    <table
                      class="table"
                    >
                      <tbody>
                        <tr>
                          <td
                            id="response-giver"
                          >
                            <strong>
                              From:
                            </strong>
                             giver1 
                          </td>
                        </tr>
                        <tr>
                          <td>
                            <tm-single-response>
                            </tm-single-response>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </tm-student-view-responses>
            </div>
          </div>
        </div>
      </div>
    </tm-loading-retry>
  </div>
</tm-question-response-panel>
`;
